internal/era2/builder2.go

```diff
@@ -0,0 +1,321 @@
+		}
+		if n, err := b.w.Write(TypeAccumulatorRoot, accRoot[:]); err != nil {
+			return common.Hash{}, fmt.Errorf("error writing accumulator root: %w", err)
+		} else {
+			b.writtenBytes += uint64(n)
+		}
+	}
+	return accRoot, b.writeIndex()
+}
+
+func uint256LE(v *big.Int) []byte {
+	b := v.FillBytes(make([]byte, 32))
+	for i := 0; i < 16; i++ {
+		b[i], b[31-i] = b[31-i], b[i]
+	}
+	return b
+}
+
+func decodeBigs(raw [][]byte) []*big.Int {
+	out := make([]*big.Int, len(raw))
+	for i, le := range raw {
+		be := make([]byte, 32)
+		for j := 0; j < 32; j++ {
+			be[j] = le[31-j]
+		}
+		out[i] = new(big.Int).SetBytes(be)
+	}
+	return out
+}
+
+func (b *Builder) snappyWrite(typ uint16, in []byte) error {
+	var (
+		buf = b.buf
+		s   = b.sn
+	)
+	buf.Reset()
+	s.Reset(buf)
+	if _, err := b.sn.Write(in); err != nil {
+		return fmt.Errorf("error snappy encoding: %w", err)
+	}
+	if err := s.Flush(); err != nil {
+		return fmt.Errorf("error flushing snappy encoding: %w", err)
+	}
+	n, err := b.w.Write(typ, b.buf.Bytes())
+	b.writtenBytes += uint64(n)
+	if err != nil {
+		return fmt.Errorf("error writing e2store entry: %w", err)
+	}
+	return nil
+}
+
+func (b *Builder) addEntry(typ uint16, payload []byte, snappyIt bool) (uint64, error) {
+	offset := b.writtenBytes
+	var err error
+	if snappyIt {
+		if err = b.snappyWrite(typ, payload); err != nil {
+			return 0, err
+		}
+	} else {
+		var n int
+		if n, err = b.w.Write(typ, payload); err != nil {
+			return 0, err
+		}
+		b.writtenBytes += uint64(n)
+	}
+	return offset, nil
+}
+
+func (b *Builder) flushKind(typ uint16, list [][]byte, useSnappy bool, dst *[]uint64) error {
+	for _, data := range list {
+		off, err := b.addEntry(typ, data, useSnappy)
+		if err != nil {
+			return fmt.Errorf("entry type %d: %w", typ, err)
+		}
+		*dst = append(*dst, off)
+	}
+	return nil
+}
+
+func (b *Builder) writeIndex() error {
+	count := uint64(len(b.headeroffsets))
+	compcount := uint64(3)
+	if len(b.tds) > 0 {
+		compcount++
+	}
+	if len(b.proofsRLP) > 0 {
+		compcount++
+	}
+
+	idx := make([]byte, 8+count*8*compcount+16) //8 for start block, 8 per property per block, 16 for the number of properties and the number of blocks
+	binary.LittleEndian.PutUint64(idx, *b.startNum)
+	base := int64(b.writtenBytes)
+	pos := 8
+	rel := func(abs uint64) uint64 { return uint64(int64(abs) - base) }
+	for i := uint64(0); i < count; i++ {
+		binary.LittleEndian.PutUint64(idx[pos:], rel(b.headeroffsets[i]))
+		pos += 8
+		binary.LittleEndian.PutUint64(idx[pos:], rel(b.bodyoffsets[i]))
+		pos += 8
+		binary.LittleEndian.PutUint64(idx[pos:], rel(b.receiptoffsets[i]))
+		pos += 8
+		if len(b.tds) > 0 {
+			binary.LittleEndian.PutUint64(idx[pos:], rel(b.tdoff[i]))
+			pos += 8
+		}
+		if len(b.proofsRLP) > 0 {
+			binary.LittleEndian.PutUint64(idx[pos:], rel(b.proofoffsets[i]))
+			pos += 8
+		}
+	}
+
+	binary.LittleEndian.PutUint64(idx[pos:], compcount)
+	pos += 8
+	binary.LittleEndian.PutUint64(idx[pos:], count)
```

Comment:
I'd drop `pos` and just compute the exact location these need to be written because it helps ensure there aren't subtle bugs in the positioning and the index matches the expected locations.